File: java/src/main/java/neetcode/slidingwindow/p0003_longestsubstringwithoutrepeatingcharacters/Solution.java
package neetcode.slidingwindow.p0003_longestsubstringwithoutrepeatingcharacters;

import java.util.HashSet;
import java.util.Set;

public class Solution {
    /**
     * Returns the length of the longest substring of s that contains no
     * duplicate characters.
     *
     * Approach: sliding window with a HashSet mirror of the current window.
     * - Expand right by one character per iteration.
     * - When a duplicate enters, shrink from the left until the duplicate
     *   is removed, then add the new character.
     * - Track the maximum window length along the way.
     *
     * Time:  O(n) — each character is added and removed at most once
     * Space: O(min(n, alphabet))
     */
    public int lengthOfLongestSubstring(String s) {
        Set<Character> window = new HashSet<>();
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);

            // Shrink from the left until c is no longer in the window
            while (window.contains(c)) {
                window.remove(s.charAt(left));
                left++;
            }

            window.add(c);

            int currentLen = right - left + 1;
            if (currentLen > maxLen) {
                maxLen = currentLen;
            }
        }

        return maxLen;
    }
}

File: python/src/neetcode/slidingwindow/p0003_longestsubstringwithoutrepeatingcharacters/solution.py
class SolutionSet:
    """
    Sliding window with a set mirroring the current window contents.
    When a duplicate is found, shrink from the left until it disappears.

    Time:  O(n) — each character added and removed at most once
    Space: O(min(n, alphabet))
    """
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = set()
        left = 0
        maxLen = 0

        for right in range(len(s)):
            while s[right] in window:
                window.remove(s[left])
                left += 1
            window.add(s[right])
            maxLen = max(maxLen, right - left + 1)

        return maxLen


class Solution:
    """
    Sliding window with a dict storing the last seen index of each character.
    Jumps left directly to lastIndex+1 instead of removing one by one.

    Time:  O(n)
    Space: O(min(n, alphabet))
    """
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = {}
        left = 0
        maxLen = 0

        for right, c in enumerate(s):
            if c in seen and seen[c] >= left:
                left = seen[c] + 1      # jump past the duplicate
            seen[c] = right
            maxLen = max(maxLen, right - left + 1)

        return maxLen
